### PR TITLE
[init] Disable GUI elements in early-init as well

### PR DESCRIPTION
**Pull request by syl20bnr**
Can provide a smoother startup by reducing graphical artifacts.

### CODE REVIEW

File: core/core-early-funcs.el [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+;;
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+(defun spacemacs/removes-gui-elements ()
+  "Remove the menu bar, tool bar and scroll bars."
+  ;; removes the GUI elements
+  (when (and (fboundp 'scroll-bar-mode) (not (eq scroll-bar-mode -1)))
```
Comment (lebensterben) [inline core/core-early-funcs.el:28]: From docstring
> To set this variable in a Lisp program, use set-scroll-bar-mode
to make it take real effect.

**Review by smile13241324 — APPROVED** [review]
Looks good.

**Comment by cjf-666** [comment]
I've discovered that disabling GUI elements in early-init causes the `green button` on title bar of Emacs app on macOS to turn into `zoom button`, instead of `full screen button`. This can be annoying obviously.
The root cause of this is `window-system` function will return `nil` when emacs is in early-init,

> (unless (memq (window-system) '(mac ns))
>    (when (and (fboundp 'menu-bar-mode) (not (eq menu-bar-mode -1)))
>      (menu-bar-mode -1)))

and the intention was to avoid disabling `menu-bar-mode` for `mac` and `ns` window system, but this can't be done in early-init stage using `window-system` function, guessing it'd be better to allow user to configure whether to disable GUI elements in early-init stage.

@syl20bnr

**Comment by lebensterben** [comment]
@cjf-666 

By `zoom button`, do you mean maximizing?

Also, can you try <kbd>⌘ cmd</kbd>-<kbd>⌃ ctrl</kbd>-<kbd>F</kbd>? That should be default keyboard shortcut for entering fullscreen mode.

**Comment by lizhuohua** [comment]
> @cjf-666
> 
> By `zoom button`, do you mean maximizing?
> 
> Also, can you try ⌘ cmd-⌃ ctrl-F? That should be default keyboard shortcut for entering fullscreen mode.

I have the same issue. The green button becomes `zoom`:

image

Instead of `fullscreen`:

image

Luckily, `command-ctrl-F` still works.

**Comment by lebensterben** [comment]
okay, so it doesn't break the functionality but only the UI.

**Comment by cjf-666** [comment]
> okay, so it doesn't break the functionality but only the UI.

Actually `command-ctrl-F` works differently from the `full screen button`, the former only `maximizing` Emacs on desktop, but the latter makes full screen out of desktop. Also the `full screen button` is the default since OSX Yosemite, I believe it is a good idea to keep `full screen button` as all other apps do.

Plus, hitting `command-ctrl-F` twice (i.e. maximizing followed by exiting maximizing) can make `zoom button` become `full screen button`, weirdly, but it can change back to `zoom button` if clicked twice (i.e. full screen followed by exiting full screen).

**Comment by lebensterben** [comment]
You can re-enable one or more of the mode in your user-config to figure out what breaks the it.

And then you can report it to upstream.

It's not a Spacemacs bug.

**Comment by cjf-666** [comment]
> It's not a Spacemacs bug.

Actually it is a Spacemacs bug, as I mentioned at the beginning,

> The root cause of this is window-system function will return nil when emacs is in early-init,
> 
>     (unless (memq (window-system) '(mac ns))
>       (when (and (fboundp 'menu-bar-mode) (not (eq menu-bar-mode -1)))
>         (menu-bar-mode -1)))
> 
> and the intention was to avoid disabling menu-bar-mode for mac and ns window system, but this can't be done in early-init stage using window-system function

In other words, commenting out `(spacemacs/removes-gui-elements)` in early-init.el will fix this problem.

**Comment by lebensterben** [comment]
Sorry I missed this part.
You're right.

@syl20bnr 
According to https://www.gnu.org/software/emacs/manual/html_node/emacs/Early-Init-File.html

> the early init file is read before the GUI is initialized, so customizations related to GUI features will not work reliably in early-init.el. By contrast, the normal init files are read after the GUI is initialized. If you must have customizations in the early init file that rely on GUI features, make them run off hooks provided by the Emacs startup, such as window-setup-hook or tty-setup-hook.

We should remove GUI elements via `window-setup-hook`.

**Comment by cjf-666** [comment]
@lebensterben Hi, I have created a PR(https://github.com/syl20bnr/spacemacs/pull/15004) fixing this, thanks for the emacs doc reference

**Pull request merged**